### PR TITLE
Move implementation from enum to Walker trait

### PR DESCRIPTION
**Pull request by AnderEnder**
* created  `Walker` trait
* moved implementations into different submodule to decrease intermodule collisions
* `WalkerConst` was moved to `test_utils`

### CODE REVIEW

File: diamond-api/src/find.rs [hunk 1]
```diff
@@ -137,174 +129,35 @@ impl FindPath {
...
-    MetricResponseLeaf {
-        name: name.to_owned(),
-        path,
-        is_leaf,
-    }
-}
+pub async fn find_handler<T: Walker>(
+    ctx: Data<Context<T>>,
+    query: FindQuery,
+) -> Result<HttpResponse, Error> {
```
Comment (andy128k) [inline diamond-api/src/find.rs]: `actix_web::Result`

File: diamond-api/src/storage.rs [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+    let full_path = dir.canonicalize()?.join(&subdir);
+    let dir_metric = subdir
+        .components()
+        .filter_map(|x| x.as_os_str().to_str())
+        .collect::<Vec<&str>>()
+        .concat();
+
+    let mut metrics: Vec<MetricResponseLeaf> = fs::read_dir(&full_path)?
+        .filter_map(|entry| {
+            let (local_path, local_file_type) = match entry {
```
Comment (andy128k) [inline diamond-api/src/storage.rs]: `
if let Ok(...)`

File: diamond-api/src/storage.rs [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+        .fold(PathBuf::new(), |acc, x| acc.join(x))
+        .with_extension("wsp");
+    let full_path = dir
+        .canonicalize()
+        .map_err(|_| ResponseError::Path)?
+        .join(&path);
+    Ok(full_path)
+}
+
+fn create_leaf(name: &str, dir: &str, is_leaf: bool) -> MetricResponseLeaf {
```
Comment (andy128k) [inline diamond-api/src/storage.rs]:
```
impl MetricResponseLeaf {
    fn new(...) -> Self { ... }
}
```

File: diamond-api/src/storage.rs [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    MetricResponseLeaf {
+        name: name.to_owned(),
+        path,
+        is_leaf,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use tempfile;
```
Comment (andy128k) [inline diamond-api/src/storage.rs]: this can be removed.

File: diamond-api/src/storage.rs [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        metrics.sort_by_key(|k| k.name.clone());
+        Ok(metrics)
+    }
+}
+
+#[inline]
+fn path(dir: &Path, metric: &str) -> Result<PathBuf, ResponseError> {
+    let path = metric
+        .split('.')
+        .fold(PathBuf::new(), |acc, x| acc.join(x))
```
Comment (andy128k) [inline diamond-api/src/storage.rs]: :+1:

File: diamond-api/src/storage.rs [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+        let full_path = self.0.canonicalize()?.join(&subdir);
+        let dir_metric = subdir
+            .components()
+            .filter_map(|x| x.as_os_str().to_str())
+            .collect::<Vec<&str>>()
+            .concat();
+
+        let mut metrics: Vec<MetricResponseLeaf> = fs::read_dir(&full_path)?
+            .filter_map(|entry| {
+                let (local_path, local_file_type) = match entry {
```
Comment (andy128k) [inline diamond-api/src/storage.rs]: `entry?`

File: diamond-api/src/storage.rs [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        &self,
+        subdir: &Path,
+        pattern: &Pattern,
+    ) -> Result<Vec<MetricResponseLeaf>, Error> {
+        let full_path = self.0.canonicalize()?.join(&subdir);
+        let dir_metric = subdir
+            .components()
+            .filter_map(|x| x.as_os_str().to_str())
+            .collect::<Vec<&str>>()
+            .concat();
```
Comment (andy128k) [inline diamond-api/src/storage.rs]: I have a feeling that `fold` will look better here. Similar to what you did with `PathBuf` below but with a `String`.

File: diamond-api/src/find.rs [hunk 8]
```diff
@@ -137,174 +129,33 @@ impl FindPath {
...
-            let metrics_completer = MetricResponse { metrics };
-            HttpResponse::Ok().json(metrics_completer)
-        }
-    })
 }
 
 #[cfg(test)]
 mod tests {
+    use actix_web::test::TestRequest;
     use serde_urlencoded;
```
Comment (andy128k) [inline diamond-api/src/find.rs]: this is redundant

File: diamond-api/src/storage.rs [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+        let dir = get_temp_dir();
+        let path = dir.path();
+        let path1 = path.join(Path::new("foo"));
+        let path2 = path.join(Path::new("bar"));
+        let path3 = path.join(Path::new("foobar.wsp"));
+        let path4 = path1.join(Path::new("bar.wsp"));
+
+        create_dir(&path1)?;
+        create_dir(&path2)?;
+        let _file1 = File::create(&path3).unwrap();
```
Comment (andy128k) [inline diamond-api/src/storage.rs]: unwrap -> ?

File: diamond-api/src/storage.rs [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+        let _file2 = File::create(&path4).unwrap();
+
+        let metric =
+            WalkerPath(path.to_owned()).walk_tree(&Path::new(""), &Pattern::new("*").unwrap())?;
```
Comment (andy128k) [inline diamond-api/src/storage.rs]: unwrap -> ?

File: diamond-api/src/storage.rs [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+                path: "foobar".into(),
+                is_leaf: true,
+            },
+        ];
+
+        metric_cmp.sort_by_key(|k| k.name.clone());
+        assert_eq!(metric, metric_cmp);
+
+        let metric2 = WalkerPath(path.to_owned())
+            .walk_tree(&Path::new("foo"), &Pattern::new("*").unwrap())?;
```
Comment (andy128k) [inline diamond-api/src/storage.rs]: unwrap -> ?

File: diamond-api/src/storage.rs (hunk 5 above)
Comment (AnderEnder) [reply diamond-api/src/storage.rs]: this is old code :)

**Pull request merged**